evm/deploy-gateway-v5.0.x.js

```diff
@@ -341,10 +366,63 @@ async function upgrade(config, options) {
         if (anwser !== 'y') return;
     }
 
-    const tx = await gateway.upgrade(contractConfig.implementation, implementationCodehash, setupParams, gasOptions);
-    printInfo('Upgrade transaction', tx.hash);
+    // Offline signing
+    let gasLimit, gasPrice;
 
-    await tx.wait(chain.confirmations);
+    try {
+        gasPrice = parseUnits((await provider.getGasPrice()).toString(), 'gwei');
+        const block = await provider.getBlock('latest');
+        gasLimit = block.gasLimit.toNumber() / 1000;
+
+        printInfo('Gas Price:', gasPrice.toString());
+        printInfo('Gas Limit:', gasLimit.toString());
+    } catch (error) {
+        printError('Gas price and limit could not be fetched from provider');
+        printObj(error);
+    }
+
+    let nonce = await getNonceFromProvider(provider, signerAddress);
+    let signersData, transactions;
+
+    if (isOffline) {
+        directoryPath = directoryPath || './txs';
+        fileName = fileName || env.toLowerCase() + '-' + chain.name.toLowerCase() + '-' + 'signedUpgradeTransactions';
+        printInfo(`Storing signed Txs offline in file ${fileName}`);
+        nonce = await getLatestNonceAndUpdateData(directoryPath, fileName, wallet);
+        signersData = await getAllSignersData(directoryPath, fileName);
+        transactions = await getTransactions(directoryPath, fileName, signerAddress);
+        const { baseTx, signedTx } = await ledgerSign(
+            gasLimit,
+            gasPrice,
+            nonce,
+            chain,
+            wallet,
+            gateway.address,
+            undefined,
+            gateway,
+            'upgrade',
+            contractConfig.implementation,
+            implementationCodehash,
+            setupParams,
+        );
+        const tx = {};
+        tx.nonce = nonce;
+        tx.msg = `This transaction will perform upgrade of AxelarGateway contract having address ${gateway.address} with implementation ${contractConfig.implementation} on chain ${chain.name} with chainId ${chain.chainId}`;
+        tx.baseTx = baseTx;
+        tx.signedTx = signedTx;
+        tx.status = 'PENDING';
+        transactions.push(tx);
+
+        if (transactions) {
+            signersData[signerAddress] = transactions;
+            await updateSignersData(directoryPath, fileName, signersData);
```

Comment:
don't think we need directory management in the script itself. we can write the tx to a provided file path. some bash scripts that sit one level higher can handle packaging etc.